src/topology/path_connected.lean

```diff
@@ -0,0 +1,589 @@
+* `is_path_connected F` asserts that `F` is non-empty and every two
+  points of `F` are joined in `F`.
+* `loc_path_connected_space X` is a predicate class asserting that `X` is locally path-connected:
+  each point has a basis of path-connected neighborhoods (we do *not* ask these to be open).
+
+## Main theorems
+
+* `joined` and `joined_in F` are transitive relations.
+
+One can link the absolute and relative version in two direction, using `(univ : set X)` or the
+subtype `↥F`.
+
+* `path_connected_space_iff_univ : path_connected_space X ↔ is_path_connected (univ : set X)`
+* `is_path_connected_iff_path_connected_space : is_path_connected F ↔ path_connected_space ↥F`
+
+For locally path connected spaces, we have
+* `path_connected_space_iff_connected_space : path_connected_space X ↔ connected_space X`
+* `is_connected_iff_is_path_connected (U_op : is_open U) : is_path_connected  U ↔ is_connected U`
+
+## Implementation notes
+
+By default, all paths have `I` as their source and `X` as their target, but there is an
+operation `I_extend` that will extend any continuous map `γ : I → X` into a continuous map
+`I_extend γ : ℝ → X` that is constant before `0` and after `1`.
+
+This is used to define `joined.extend` that turns `h : joined x y` into a continuous map
+`h.extend : ℝ → X` whose restriction to `I` joins `x` and `y`.
+
+Similarly, one can turn `h : joined_in F x y` into a continuous map
+`h.extend : ℝ → X`, and `h.map : I → F` taking values in the subtype `F`,
+and also `h.extend_map : ℝ → F`.
+
+-/
+
+noncomputable theory
+open_locale classical topological_space filter
+open filter set
+
+variables {X : Type*} [topological_space X] {x y z : X} {ι : Type*}
+
+local notation `I` := Icc (0 : ℝ) 1
+
+lemma Icc_zero_one_symm {t : ℝ} : t ∈ I ↔ 1 - t ∈ I :=
+begin
+  rw [mem_Icc, mem_Icc],
+  split ; intro ; split ; linarith
+end
+
+instance I_has_zero : has_zero I := ⟨⟨0, by split ; norm_num⟩⟩
+
+@[simp, norm_cast] lemma coe_I_zero : ((0 : I) : ℝ) = 0 := rfl
+
+instance I_has_one : has_one I := ⟨⟨1, by split ; norm_num⟩⟩
+
+@[simp, norm_cast] lemma coe_I_one : ((1 : I) : ℝ) = 1 := rfl
+
+/-- Unit interval central symmetry. -/
+def I_symm : I → I := λ t, ⟨1 - t.val, Icc_zero_one_symm.mp t.property⟩
+
+local notation `σ` := I_symm
+
+@[simp] lemma I_symm_zero : σ 0 = 1 :=
+subtype.ext $ by simp [I_symm]
+
+@[simp] lemma I_symm_one : σ 1 = 0 :=
+subtype.ext $ by simp [I_symm]
+
+@[continuity]
+lemma continuous_I_symm : continuous σ :=
+by continuity!
+
+/-- Projection of `ℝ` onto its unit interval. -/
+def proj_I : ℝ → I :=
+λ t, if h : t ≤ 0 then ⟨0, left_mem_Icc.mpr zero_le_one⟩ else
+     if h' : t ≤ 1 then ⟨t, ⟨le_of_lt $ not_le.mp h, h'⟩⟩ else ⟨1, right_mem_Icc.mpr zero_le_one⟩
+
+lemma proj_I_I {t : ℝ} (h : t ∈ I) : proj_I t = ⟨t, h⟩ :=
+begin
+  unfold proj_I,
+  rw mem_Icc at h,
+  split_ifs,
+  { simp [show t = 0, by linarith] },
+  { refl },
+  { exfalso, linarith }
+end
+
+lemma range_proj_I : range proj_I = univ :=
```

Comment:
I'd add `surjective_proj_I : surjective proj_I`. It gives you at least [range_eq](https://leanprover-community.github.io/mathlib_docs/data/set/basic.html#function.surjective.range_eq) and [range_comp](https://leanprover-community.github.io/mathlib_docs/data/set/basic.html#function.surjective.range_comp) (`I_extend_range` below).